### PR TITLE
Feature/base sampling

### PR DESCRIPTION
**Pull request by rikba**
This PR introduces (automatic) base station sampling using a least square estimate. 
- The sampling can be initiated using a service call or automatically through rosparam
- The sampling will average N covariance weighted ECEF samples using a least squares estimator
- The ECEF position is converted to WGS84 using geotf and set in the piksi firmware
- Additionally any receiver type can sample its static position using the least squares estimator initiated by a service call

To be reviewed after https://github.com/ethz-asl/ethz_piksi_ros/pull/124. Adresses base sampling capabilities mentioned in https://github.com/ethz-asl/ethz_piksi_ros/issues/102.

This code may have a "code smell" as the base station receiver is subscribing to its own ROS topic to indicate when the sampling has finished.

### CODE REVIEW

**Comment by michaelpantic** [comment]
Can you merge new master? Then ill have a look ;-)

**Comment by rikba** [comment]
Done @michaelpantic

File: piksi_multi_cpp/src/sbp_callback_handler/position_sampler.cc [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+  lrm::convertCartesianPoint<msg_pos_ecef_cov_t>(*sbp_msg, &z);
+  Eigen::Matrix3d R;
+  lrm::convertCartesianCov<msg_pos_ecef_cov_t>(*sbp_msg, &R);
+
+  // Cache least square measurement values.
+  size_t block_idx = 3 * num_fixes_++;
+  y_.value().segment(block_idx, 3) = z;
+  R_inv_.value().block<3, 3>(block_idx, block_idx) = R.inverse();
+
+  if (x_.has_value() && P_.has_value()) {
```
Comment (michaelpantic) [inline piksi_multi_cpp/src/sbp_callback_handler/position_sampler.cc]: add comment that this is the kalman update

**Comment by michaelpantic** [comment]
> This code may have a "code smell" as the base station receiver is subscribing to its own ROS topic to indicate when the sampling has finished.

Hmm, that's a bit ugly but if its documented nicely somewhere that should be fine. It has the advantage that the sampler node also could be something external if needed.

BTW, how do you decide if sampling should be run at all?

**Comment by rikba** [comment]
> BTW, how do you decide if sampling should be run at all?

First of all every receiver type can sample its static position using the service call `sample_position`. The base station additionally has a service call to resample the position and write it to the settings `overwrite_position` (should probably rename). And the base station can automatically sample its position when setting a rosparam.

**Comment by rikba** [comment]
Thanks for the review. I documented the code behavior and improved the sampling interface.

**Comment by michaelpantic** [comment]
> > BTW, how do you decide if sampling should be run at all?
> 
> First of all every receiver type can sample its static position using the service call `sample_position`. The base station additionally has a service call to resample the position and write it to the settings `overwrite_position` (should probably rename). And the base station can automatically sample its position when setting a rosparam.

Awesome, very nice functionality.. thanks for the clarification!

**Pull request merged**